rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,68 @@
+- Other infrastructure as code tools in use (e.g. [Fabric](https://github.com/alphagov/fabric-scripts)) are also currently using very old versions which are very likely to have problems with newer operating systems.
+
+## Proposal
+
+GOV.UK will use a managed, container based hosting environment, such as [GOV.UK PaaS](https://www.cloud.service.gov.uk). The choice of hosting environment is out of scope for this RFC.
+
+GOV.UK applications and their dependencies will be built into container images. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.
+
+A containerised runtime aligns with the [GOV.UK Infrastructure Architecture Goals for 2021](
+https://docs.google.com/document/d/1ooN7wkYhEGvceGe9Qz_HNZa-GPtrjzK_vA4vfWYVn4c/edit#heading=h.cdrr7rv9t98f) to isolate applications through containers. This will give us better control of how resources are matched to applications.
+
+The TechOps strategy is to use common components where available. GOV.UK has been following the strategy by using hosted versions of software ([ElasticSearch](https://aws.amazon.com/elasticsearch-service/) and [Postgres](https://aws.amazon.com/rds/)) and services ([Notify](https://www.notifications.service.gov.uk)) where available. Using a containerised runtime is a continuation of this policy.
+
+GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.
+
+### Benefits of this approach
+
+The most immediate benefit of this approach is that it will allow us to incrementally upgrade the operating system version. By moving the applications into containers, GOV.UK can enable a much lower risk upgrade path.
+
+Moving to a container based hosting environment will allow us to remove much of our legacy Puppet and Fabric code. This code is hard to maintain, and hard to hire people with the experience required to improve and upgrade. With industry and government overwhelmingly moving towards containerised platforms, it should be much easier to hire people with these skills.
+
+Containerised hosting environments can generally provide much better value for money in terms of resource usage, both because many small applications can be “bin-packed” onto the same infrastructure, and because it is much easier to autoscale applications.
+
+Running containers in production allows for a reduction in the difference between development environments and production. This can lead to bugs being found more quickly, and faster development cycles.
+
+Other improvements to GOV.UK’s development lifecycle (such as moving to continuous deployment) will be easier to implement on a container based infrastructure than they would in the current virtual machine based infrastructure.
+
+### Risks of this approach
```

Comment:
I think the reason we have so much outdated infra (OS, puppet, fabric, etc) is because it works well enough for us to not prioritise updating it.  And now it's become a big problem.

What guarantee do we have that any container stuff we implement won't become similarly outdated?

Reply:
I don't think we have a guarantee, but if we can offload the need to support the OS and related dependencies to the platform provider then it should give us more time to keep any container stuff up to date.

Reply:
I think that one of the benefits of using containers is that it breaks the big problems down in to smaller ones.

The OS, fabric and puppet problems are all big because they touch so many things, which creates inertia to address the problem, and then deprioritisation.

Compare it with containers:  lets say SomeLib needs updating from v2 to v3 in 20 containerised apps.  We can now break it in to 20 smaller pieces of work with different priorities; making the problem smaller also means we can have higher confidence and lower impact from things like dependabot auto-upgrading libraries and auto feeding that auto-upgrade in to the testing pipeline.

Where I am going with this all is: i think containers will make the work smaller, easier, routine and possibly even dissapear if we can automate it enough.  Meaning prioritisation will become less of a problem and help mitigate this problem appearing again

Reply:
Added these sentences to the RFC in 101523a :

> Future upgrades should be easier to prioritise, because they will be smaller, lower risk pieces of work.

> GOV.UK will still need to continuously upgrade some infrastructure dependencies. There is a risk that this work will not be prioritised, leaving GOV.UK with a different set of out-of-date infrastructure in the future. Moving to managed platforms wherever possible helps mitigate this, but there will always be some maintenance work that needs to be prioritised.